{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d4e6a0a1",
   "metadata": {},
   "source": [
    "# Relative quantification proof of concept\n",
    "\n",
    "Take a first pass at implementing evaluation of (fractional) relative quantification, more or less following [Joseph's blueprint](https://docs.google.com/document/d/1u18BSWB2776d-EXAoft7Je5KmhC1aGMFZJRpJ1rueNI/edit) with a [few practical pointers I had come up with](https://docs.google.com/document/d/1t30SEHZ6RSFYXG7o1ylWCfgYWVbZCA1jj-SkES1M2rw/edit)\n",
    "\n",
    "Purpose isn't to necessarily to make something fully flexible/ready to go, but to explore how to implement the blueprint and understand any impacts our filtering criteria/decisions may have.\n",
    "\n",
    "\n",
    "Broadly, this involves:\n",
    "- Defining list of genes expressed above a given TPM threshold from RNA-seq data\n",
    "- Defining a set of terminal exons w/ 2 polyA sites from the GT poly(A) site BEDs, filtering to 2 most expressed sites if a TE has > 2 PAS\n",
    "- Filtering the predicted sites to those occurring on curated set of terminal exons\n",
    "- \n",
    "\n",
    "\n",
    "Will focus on a single Mayr sample Mayr_CD5B_R3 with outputs from DaPars2 & DaPars for now..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d0d3064",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import warnings\n",
    "\n",
    "# There are some FutureWarnings with installed version of PyRanges and Pandas - For visual purposes just don't print them for now\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5be1dc70",
   "metadata": {},
   "source": [
    "# 1. Find genes expressed in RNA-seq above a given threshold\n",
    "\n",
    "\n",
    "Matt provided a TSV of Salmon quantification outputs from the nf-core RNA-seq workflow (**TODO: insert link**). As this is per-transcript, need to sum to gene-level\n",
    "\n",
    "Target is a list/set of Ensembl gene IDs that pass a given TPM threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "50fb0449",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Mayr_CD5B_R3</th>\n",
       "      <th>Mayr_CD5B_R4</th>\n",
       "      <th>Mayr_GC_R1</th>\n",
       "      <th>Mayr_GC_R2</th>\n",
       "      <th>Mayr_GC_R3</th>\n",
       "      <th>Mayr_GC_R4</th>\n",
       "      <th>Mayr_M_R2</th>\n",
       "      <th>Mayr_M_R6</th>\n",
       "      <th>Mayr_NB_R1</th>\n",
       "      <th>Mayr_NB_R2</th>\n",
       "      <th>Mayr_NB_R3</th>\n",
       "      <th>Mayr_NB_R4</th>\n",
       "      <th>Mayr_NB_R5</th>\n",
       "      <th>Mayr_NB_R6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.044865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.135770</td>\n",
       "      <td>0.012529</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>2.62799</td>\n",
       "      <td>3.314172</td>\n",
       "      <td>2.877079</td>\n",
       "      <td>2.625817</td>\n",
       "      <td>2.839663</td>\n",
       "      <td>3.452287</td>\n",
       "      <td>2.285098</td>\n",
       "      <td>2.630034</td>\n",
       "      <td>2.627121</td>\n",
       "      <td>2.015704</td>\n",
       "      <td>2.529293</td>\n",
       "      <td>1.699014</td>\n",
       "      <td>2.747853</td>\n",
       "      <td>2.22484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.03989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026530</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>ENSG00000268020.3</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.073951</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tx            gene_id  Mayr_CD5B_R3  Mayr_CD5B_R4  \\\n",
       "0  ENST00000456328.2  ENSG00000223972.5       0.00000      0.044865   \n",
       "1  ENST00000450305.2  ENSG00000223972.5       0.00000      0.000000   \n",
       "2  ENST00000488147.1  ENSG00000227232.5       2.62799      3.314172   \n",
       "3  ENST00000619216.1  ENSG00000278267.1       0.00000      0.000000   \n",
       "4  ENST00000473358.1  ENSG00000243485.5       0.00000      0.000000   \n",
       "5  ENST00000469289.1  ENSG00000243485.5       0.00000      0.000000   \n",
       "6  ENST00000607096.1  ENSG00000284332.1       0.00000      0.000000   \n",
       "7  ENST00000417324.1  ENSG00000237613.2       0.00000      0.000000   \n",
       "8  ENST00000461467.1  ENSG00000237613.2       0.00000      0.000000   \n",
       "9  ENST00000606857.1  ENSG00000268020.3       0.00000      0.000000   \n",
       "\n",
       "   Mayr_GC_R1  Mayr_GC_R2  Mayr_GC_R3  Mayr_GC_R4  Mayr_M_R2  Mayr_M_R6  \\\n",
       "0    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "1    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "2    2.877079    2.625817    2.839663    3.452287   2.285098   2.630034   \n",
       "3    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "4    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "5    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "6    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "7    0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   \n",
       "8    0.000000    0.026530    0.000000    0.000000   0.030134   0.000000   \n",
       "9    0.000000    0.000000    0.000000    0.000000   0.073951   0.000000   \n",
       "\n",
       "   Mayr_NB_R1  Mayr_NB_R2  Mayr_NB_R3  Mayr_NB_R4  Mayr_NB_R5  Mayr_NB_R6  \n",
       "0    0.135770    0.012529    0.000000    0.000000    0.000000     0.00000  \n",
       "1    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  \n",
       "2    2.627121    2.015704    2.529293    1.699014    2.747853     2.22484  \n",
       "3    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  \n",
       "4    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  \n",
       "5    0.000000    0.000000    0.000000    0.000000    0.000000     0.03989  \n",
       "6    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  \n",
       "7    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  \n",
       "8    0.000000    0.000000    0.026825    0.000000    0.000000     0.00000  \n",
       "9    0.000000    0.000000    0.000000    0.000000    0.000000     0.00000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_df = pd.read_csv(\"data/Mayr_salmon.merged.transcript_tpm.tsv\", sep=\"\\t\")\n",
    "tpm_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "743e85b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mayr_CD5B_R3</th>\n",
       "      <th>Mayr_CD5B_R4</th>\n",
       "      <th>Mayr_GC_R1</th>\n",
       "      <th>Mayr_GC_R2</th>\n",
       "      <th>Mayr_GC_R3</th>\n",
       "      <th>Mayr_GC_R4</th>\n",
       "      <th>Mayr_M_R2</th>\n",
       "      <th>Mayr_M_R6</th>\n",
       "      <th>Mayr_NB_R1</th>\n",
       "      <th>Mayr_NB_R2</th>\n",
       "      <th>Mayr_NB_R3</th>\n",
       "      <th>Mayr_NB_R4</th>\n",
       "      <th>Mayr_NB_R5</th>\n",
       "      <th>Mayr_NB_R6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>0.841496</td>\n",
       "      <td>0.313483</td>\n",
       "      <td>0.224971</td>\n",
       "      <td>0.220765</td>\n",
       "      <td>0.022432</td>\n",
       "      <td>0.569883</td>\n",
       "      <td>0.789282</td>\n",
       "      <td>0.378343</td>\n",
       "      <td>0.021219</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.269005</td>\n",
       "      <td>0.835454</td>\n",
       "      <td>0.435978</td>\n",
       "      <td>0.007870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.053663</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.14</th>\n",
       "      <td>31.462058</td>\n",
       "      <td>27.354208</td>\n",
       "      <td>35.251502</td>\n",
       "      <td>29.225256</td>\n",
       "      <td>24.685265</td>\n",
       "      <td>34.762234</td>\n",
       "      <td>28.490978</td>\n",
       "      <td>18.881510</td>\n",
       "      <td>35.288736</td>\n",
       "      <td>26.330158</td>\n",
       "      <td>24.401633</td>\n",
       "      <td>22.140657</td>\n",
       "      <td>19.653655</td>\n",
       "      <td>20.561543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>13.687977</td>\n",
       "      <td>13.619168</td>\n",
       "      <td>6.316804</td>\n",
       "      <td>5.554757</td>\n",
       "      <td>4.834584</td>\n",
       "      <td>5.143130</td>\n",
       "      <td>9.068360</td>\n",
       "      <td>5.792793</td>\n",
       "      <td>8.083517</td>\n",
       "      <td>0.864191</td>\n",
       "      <td>8.795230</td>\n",
       "      <td>7.967781</td>\n",
       "      <td>6.033806</td>\n",
       "      <td>1.219366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>6.285515</td>\n",
       "      <td>4.806302</td>\n",
       "      <td>22.109690</td>\n",
       "      <td>18.177691</td>\n",
       "      <td>14.991760</td>\n",
       "      <td>14.880926</td>\n",
       "      <td>4.207996</td>\n",
       "      <td>4.209314</td>\n",
       "      <td>1.514664</td>\n",
       "      <td>0.489746</td>\n",
       "      <td>5.427459</td>\n",
       "      <td>3.280607</td>\n",
       "      <td>3.677558</td>\n",
       "      <td>0.469440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288721.1</th>\n",
       "      <td>0.299191</td>\n",
       "      <td>0.590024</td>\n",
       "      <td>0.962898</td>\n",
       "      <td>1.188937</td>\n",
       "      <td>0.839328</td>\n",
       "      <td>1.087980</td>\n",
       "      <td>0.350958</td>\n",
       "      <td>0.304280</td>\n",
       "      <td>0.505675</td>\n",
       "      <td>0.029361</td>\n",
       "      <td>0.246273</td>\n",
       "      <td>0.191482</td>\n",
       "      <td>0.293596</td>\n",
       "      <td>0.159076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288722.1</th>\n",
       "      <td>3.003168</td>\n",
       "      <td>0.794827</td>\n",
       "      <td>4.912289</td>\n",
       "      <td>2.726699</td>\n",
       "      <td>1.875441</td>\n",
       "      <td>7.280854</td>\n",
       "      <td>0.602439</td>\n",
       "      <td>2.423762</td>\n",
       "      <td>1.573895</td>\n",
       "      <td>1.523601</td>\n",
       "      <td>4.048628</td>\n",
       "      <td>0.883793</td>\n",
       "      <td>2.006525</td>\n",
       "      <td>1.103560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288723.1</th>\n",
       "      <td>0.032072</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029104</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288724.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288725.1</th>\n",
       "      <td>1.517664</td>\n",
       "      <td>1.601146</td>\n",
       "      <td>1.266355</td>\n",
       "      <td>1.549774</td>\n",
       "      <td>1.635822</td>\n",
       "      <td>2.798908</td>\n",
       "      <td>2.485099</td>\n",
       "      <td>1.508843</td>\n",
       "      <td>0.841387</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.612817</td>\n",
       "      <td>0.855837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60649 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Mayr_CD5B_R3  Mayr_CD5B_R4  Mayr_GC_R1  Mayr_GC_R2  \\\n",
       "gene_id                                                                  \n",
       "ENSG00000000003.15      0.841496      0.313483    0.224971    0.220765   \n",
       "ENSG00000000005.6       0.000000      0.000000    0.000000    0.000000   \n",
       "ENSG00000000419.14     31.462058     27.354208   35.251502   29.225256   \n",
       "ENSG00000000457.14     13.687977     13.619168    6.316804    5.554757   \n",
       "ENSG00000000460.17      6.285515      4.806302   22.109690   18.177691   \n",
       "...                          ...           ...         ...         ...   \n",
       "ENSG00000288721.1       0.299191      0.590024    0.962898    1.188937   \n",
       "ENSG00000288722.1       3.003168      0.794827    4.912289    2.726699   \n",
       "ENSG00000288723.1       0.032072      0.000000    0.000000    0.000000   \n",
       "ENSG00000288724.1       0.000000      0.000000    0.000000    0.000000   \n",
       "ENSG00000288725.1       1.517664      1.601146    1.266355    1.549774   \n",
       "\n",
       "                    Mayr_GC_R3  Mayr_GC_R4  Mayr_M_R2  Mayr_M_R6  Mayr_NB_R1  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000000003.15    0.022432    0.569883   0.789282   0.378343    0.021219   \n",
       "ENSG00000000005.6     0.000000    0.000000   0.000000   0.000000    0.000000   \n",
       "ENSG00000000419.14   24.685265   34.762234  28.490978  18.881510   35.288736   \n",
       "ENSG00000000457.14    4.834584    5.143130   9.068360   5.792793    8.083517   \n",
       "ENSG00000000460.17   14.991760   14.880926   4.207996   4.209314    1.514664   \n",
       "...                        ...         ...        ...        ...         ...   \n",
       "ENSG00000288721.1     0.839328    1.087980   0.350958   0.304280    0.505675   \n",
       "ENSG00000288722.1     1.875441    7.280854   0.602439   2.423762    1.573895   \n",
       "ENSG00000288723.1     0.000000    0.000000   0.029104   0.000000    0.000000   \n",
       "ENSG00000288724.1     0.000000    0.000000   0.000000   0.000000    0.000000   \n",
       "ENSG00000288725.1     1.635822    2.798908   2.485099   1.508843    0.841387   \n",
       "\n",
       "                    Mayr_NB_R2  Mayr_NB_R3  Mayr_NB_R4  Mayr_NB_R5  Mayr_NB_R6  \n",
       "gene_id                                                                         \n",
       "ENSG00000000003.15    0.000000    1.269005    0.835454    0.435978    0.007870  \n",
       "ENSG00000000005.6     0.000000    0.000000    0.053663    0.000000    0.000000  \n",
       "ENSG00000000419.14   26.330158   24.401633   22.140657   19.653655   20.561543  \n",
       "ENSG00000000457.14    0.864191    8.795230    7.967781    6.033806    1.219366  \n",
       "ENSG00000000460.17    0.489746    5.427459    3.280607    3.677558    0.469440  \n",
       "...                        ...         ...         ...         ...         ...  \n",
       "ENSG00000288721.1     0.029361    0.246273    0.191482    0.293596    0.159076  \n",
       "ENSG00000288722.1     1.523601    4.048628    0.883793    2.006525    1.103560  \n",
       "ENSG00000288723.1     0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "ENSG00000288724.1     0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "ENSG00000288725.1     0.000000    0.612817    0.855837    0.000000    0.000000  \n",
       "\n",
       "[60649 rows x 14 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summarise individual transcript expression to gene-level TPM\n",
    "sample_cols = [col for col in tpm_df.columns if col not in [\"tx\", \"gene_id\"]]\n",
    "# sample_cols\n",
    "\n",
    "gene_tpm = tpm_df.groupby(\"gene_id\")[sample_cols].sum()\n",
    "gene_tpm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa33b32d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mayr_CD5B_R3</th>\n",
       "      <th>Mayr_CD5B_R4</th>\n",
       "      <th>Mayr_GC_R1</th>\n",
       "      <th>Mayr_GC_R2</th>\n",
       "      <th>Mayr_GC_R3</th>\n",
       "      <th>Mayr_GC_R4</th>\n",
       "      <th>Mayr_M_R2</th>\n",
       "      <th>Mayr_M_R6</th>\n",
       "      <th>Mayr_NB_R1</th>\n",
       "      <th>Mayr_NB_R2</th>\n",
       "      <th>Mayr_NB_R3</th>\n",
       "      <th>Mayr_NB_R4</th>\n",
       "      <th>Mayr_NB_R5</th>\n",
       "      <th>Mayr_NB_R6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.14</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288721.1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288722.1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288723.1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288724.1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288725.1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60649 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Mayr_CD5B_R3  Mayr_CD5B_R4  Mayr_GC_R1  Mayr_GC_R2  \\\n",
       "gene_id                                                                  \n",
       "ENSG00000000003.15         False         False       False       False   \n",
       "ENSG00000000005.6          False         False       False       False   \n",
       "ENSG00000000419.14          True          True        True        True   \n",
       "ENSG00000000457.14          True          True        True        True   \n",
       "ENSG00000000460.17          True          True        True        True   \n",
       "...                          ...           ...         ...         ...   \n",
       "ENSG00000288721.1          False         False       False        True   \n",
       "ENSG00000288722.1           True         False        True        True   \n",
       "ENSG00000288723.1          False         False       False       False   \n",
       "ENSG00000288724.1          False         False       False       False   \n",
       "ENSG00000288725.1           True          True        True        True   \n",
       "\n",
       "                    Mayr_GC_R3  Mayr_GC_R4  Mayr_M_R2  Mayr_M_R6  Mayr_NB_R1  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000000003.15       False       False      False      False       False   \n",
       "ENSG00000000005.6        False       False      False      False       False   \n",
       "ENSG00000000419.14        True        True       True       True        True   \n",
       "ENSG00000000457.14        True        True       True       True        True   \n",
       "ENSG00000000460.17        True        True       True       True        True   \n",
       "...                        ...         ...        ...        ...         ...   \n",
       "ENSG00000288721.1        False        True      False      False       False   \n",
       "ENSG00000288722.1         True        True      False       True        True   \n",
       "ENSG00000288723.1        False       False      False      False       False   \n",
       "ENSG00000288724.1        False       False      False      False       False   \n",
       "ENSG00000288725.1         True        True       True       True       False   \n",
       "\n",
       "                    Mayr_NB_R2  Mayr_NB_R3  Mayr_NB_R4  Mayr_NB_R5  Mayr_NB_R6  \n",
       "gene_id                                                                         \n",
       "ENSG00000000003.15       False        True       False       False       False  \n",
       "ENSG00000000005.6        False       False       False       False       False  \n",
       "ENSG00000000419.14        True        True        True        True        True  \n",
       "ENSG00000000457.14       False        True        True        True        True  \n",
       "ENSG00000000460.17       False        True        True        True       False  \n",
       "...                        ...         ...         ...         ...         ...  \n",
       "ENSG00000288721.1        False       False       False       False       False  \n",
       "ENSG00000288722.1         True        True       False        True        True  \n",
       "ENSG00000288723.1        False       False       False       False       False  \n",
       "ENSG00000288724.1        False       False       False       False       False  \n",
       "ENSG00000288725.1        False       False       False       False       False  \n",
       "\n",
       "[60649 rows x 14 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a matrix of Booleans for each column for whether gene passes expression filter\n",
    "thresh_tpm1 = gene_tpm > 1\n",
    "thresh_tpm1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f312dd23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mayr_CD5B_R3    12718\n",
       "Mayr_CD5B_R4    12288\n",
       "Mayr_GC_R1      12662\n",
       "Mayr_GC_R2      12436\n",
       "Mayr_GC_R3      11942\n",
       "Mayr_GC_R4      12673\n",
       "Mayr_M_R2       12103\n",
       "Mayr_M_R6       12321\n",
       "Mayr_NB_R1      11639\n",
       "Mayr_NB_R2       9477\n",
       "Mayr_NB_R3      12245\n",
       "Mayr_NB_R4      11996\n",
       "Mayr_NB_R5      12070\n",
       "Mayr_NB_R6       9645\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of genes passing threshold in each sample\n",
    "thresh_tpm1.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3adc2603",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mayr_CD5B_R3    0.209698\n",
       "Mayr_CD5B_R4    0.202608\n",
       "Mayr_GC_R1      0.208775\n",
       "Mayr_GC_R2      0.205049\n",
       "Mayr_GC_R3      0.196903\n",
       "Mayr_GC_R4      0.208956\n",
       "Mayr_M_R2       0.199558\n",
       "Mayr_M_R6       0.203153\n",
       "Mayr_NB_R1      0.191908\n",
       "Mayr_NB_R2      0.156260\n",
       "Mayr_NB_R3      0.201899\n",
       "Mayr_NB_R4      0.197794\n",
       "Mayr_NB_R5      0.199014\n",
       "Mayr_NB_R6      0.159030\n",
       "dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fraction of annotated genes passing threshold in each sample\n",
    "# NB: could subset to prot-coding & lncRNA genes (as expect their RNAs to be polyadenylated)\n",
    "# don't think this is that informative as don't expect all annotated genes to be expressed...\n",
    "thresh_tpm1.sum(axis=0) / len(thresh_tpm1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "51044cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dict of {sample_name: {gene_ids}} that pass the expression filter\n",
    "genes_thresh_tpm1 = {col: set(thresh_tpm1.index[thresh_tpm1.loc[:, col]]) for col in thresh_tpm1.columns}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "396c5e3b",
   "metadata": {},
   "source": [
    "# 2 - Find / Decide TEs to select for filtering ground truth\n",
    "\n",
    "- Merge/'union' overlapping TEs to create the longest possible TE isoform for searching in GT for overlapping sites (most permissive)\n",
    "    - In cases where TE isoforms have different 3'ss, they will be merged regardless (i.e. not considered as separate events)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c38989db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 14s, sys: 4.43 s, total: 1min 18s\n",
      "Wall time: 1min 19s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gtf = pr.read_gtf(\"data/gencode.v38.annotation.gtf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e62b37d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    |     Start |       End | +20   |\n",
       "| (category)   | (object)   | (object)   |   (int32) |   (int32) | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | gene       |     11868 |     14409 | ...   |\n",
       "| chr1         | HAVANA     | transcript |     11868 |     14409 | ...   |\n",
       "| chr1         | HAVANA     | exon       |     11868 |     12227 | ...   |\n",
       "| chr1         | HAVANA     | exon       |     12612 |     12721 | ...   |\n",
       "| chr1         | HAVANA     | exon       |     13220 |     14409 | ...   |\n",
       "| chr1         | HAVANA     | transcript |     12009 |     13670 | ...   |\n",
       "| chr1         | HAVANA     | exon       |     12009 |     12057 | ...   |\n",
       "| chr1         | HAVANA     | exon       |     12178 |     12227 | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 25 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "20 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 13 more.)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PyRanges objects are dictionaries of pandas dataframes\n",
    "# Keys are chrom names & strand, values are dfs storing genomic coords and metadata\n",
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b362cefc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.stranded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "679debe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>3</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150419</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150420</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150421</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>3</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150422</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>4</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150423</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3150424 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | gene       | 11868     | 14409     | +            | ...   |\n",
       "| chr1         | transcript | 11868     | 14409     | +            | ...   |\n",
       "| chr1         | exon       | 11868     | 12227     | +            | ...   |\n",
       "| chr1         | exon       | 12612     | 12721     | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 57213879  | 57213964  | -            | ...   |\n",
       "| chrY         | exon       | 57213525  | 57213602  | -            | ...   |\n",
       "| chrY         | exon       | 57213203  | 57213357  | -            | ...   |\n",
       "| chrY         | exon       | 57212183  | 57213125  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 3,150,424 rows and 12 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 1 more.)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = gtf[[\"Feature\", \"gene_id\", \"transcript_id\", \"gene_name\", \"exon_number\", \"gene_type\", \"transcript_type\", \"transcript_support_level\"]]\n",
    "\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1dbc71ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome                  category\n",
       "Feature                       object\n",
       "Start                          int32\n",
       "End                            int32\n",
       "Strand                      category\n",
       "gene_id                       object\n",
       "transcript_id                 object\n",
       "gene_name                     object\n",
       "exon_number                    int64\n",
       "gene_type                     object\n",
       "transcript_type               object\n",
       "transcript_support_level      object\n",
       "dtype: object"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For simplicities sake let's focus on a single sample\n",
    "# Mayr_CD5B_R3\n",
    "exons = gtf.subset(lambda df: df[\"Feature\"] == \"exon\")\n",
    "# Need as int dtype for some downstream functions to work\n",
    "exons = exons.assign(\"exon_number\", lambda df: df[\"exon_number\"].astype(float).astype(int))\n",
    "\n",
    "exons.dtypes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "aabebc48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>629061</td>\n",
       "      <td>629433</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000225972.1</td>\n",
       "      <td>ENST00000416931.1</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000225630.1</td>\n",
       "      <td>ENST00000457540.1</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>631073</td>\n",
       "      <td>632616</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237973.1</td>\n",
       "      <td>ENST00000414273.1</td>\n",
       "      <td>MTCO1P12</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>633695</td>\n",
       "      <td>634376</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000248527.1</td>\n",
       "      <td>ENST00000514057.1</td>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>634375</td>\n",
       "      <td>634922</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198744.5</td>\n",
       "      <td>ENST00000416718.2</td>\n",
       "      <td>MTCO3P12</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894990</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20506768</td>\n",
       "      <td>20507456</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229236.3</td>\n",
       "      <td>ENST00000666666.1</td>\n",
       "      <td>TTTY10</td>\n",
       "      <td>4</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894991</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20570221</td>\n",
       "      <td>20570519</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277438.1</td>\n",
       "      <td>ENST00000622542.1</td>\n",
       "      <td>KDM5DP1</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894992</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>21044598</td>\n",
       "      <td>21044724</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>RP11-65G9.1</td>\n",
       "      <td>1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894993</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>21042268</td>\n",
       "      <td>21042369</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>RP11-65G9.1</td>\n",
       "      <td>2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894994</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>RP11-65G9.1</td>\n",
       "      <td>3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>894995 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 629061    | 629433    | +            | ...   |\n",
       "| chr1         | exon       | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | exon       | 631073    | 632616    | +            | ...   |\n",
       "| chr1         | exon       | 633695    | 634376    | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 20570221  | 20570519  | -            | ...   |\n",
       "| chrY         | exon       | 21044598  | 21044724  | -            | ...   |\n",
       "| chrY         | exon       | 21042268  | 21042369  | -            | ...   |\n",
       "| chrY         | exon       | 21038288  | 21039044  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 894,995 rows and 12 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 1 more.)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_mayr_cd_r3_tpm_filt = exons.subset(lambda df: df[\"gene_id\"].isin(genes_thresh_tpm1[\"Mayr_CD5B_R3\"]))\n",
    "\n",
    "exons_mayr_cd_r3_tpm_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a34e4ae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract terminal exons for each transcript\n",
    "def get_terminal_exons(gr,\n",
    "                         feature_col = \"Feature\",\n",
    "                         feature_key = \"exon\",\n",
    "                         id_col = \"transcript_id\",\n",
    "                         region_number_col = \"exon_number\",\n",
    "                         filter_single = False\n",
    "                         ):\n",
    "    '''\n",
    "    Return gr of last exons for each transcript\n",
    "    \n",
    "    Function makes the following assumptions:\n",
    "    - region_number_col values define non-overlapping regions (exons) of the same group (transcript_id)\n",
    "    - region_number_col is strand aware 5'-3' in ascending order\n",
    "        - i.e. 1st exon (most 5') is numbered 1, last exon of transcript with n exons is labelled n\n",
    "        - For reference GTFs this is probably a fair assumption\n",
    "    - ONLY a single unique value (feature_key) is present in feature_col (e.g. only 'exon' present)\n",
    "    - region_number_col is an int np.dtype col\n",
    "    - regions are not duplicated within the gr\n",
    "    \n",
    "    Notes:\n",
    "    - use add_region_number if not confident region_number_col matches expectation\n",
    "        \n",
    "    '''\n",
    "\n",
    "    assert region_number_col in gr.columns.tolist()\n",
    "    assert feature_col in gr.columns.tolist()\n",
    "    assert id_col in gr.columns.tolist()\n",
    "\n",
    "    # Make sure region_number is an int dtype (so can be sorted numerically)\n",
    "    assert gr.dtypes.loc[region_number_col] in [np.dtype('int32'), np.dtype('int64')]\n",
    "    # Make sure only 'exon' features are in the gr\n",
    "    assert gr.as_df()[feature_col].drop_duplicates().tolist() == [feature_key], \"only {} entries should be present in gr\".format(feature_key)\n",
    "\n",
    "\n",
    "    # Make sure gr is sorted by transcript_id & 'region number' (ascending order so 1..n)\n",
    "    mod_gr = gr.apply(lambda df: df.sort_values(by=[id_col, region_number_col], ascending=True),\n",
    "                          nb_cpu=1)\n",
    "\n",
    "\n",
    "    # Filter out single-exon transcripts\n",
    "    if filter_single:\n",
    "        print(\"Filtering for multi-exon transcripts...\")\n",
    "        print(f\"Before: {len(set(mod_gr.as_df()[id_col]))}\")\n",
    "\n",
    "        # Setting to 'False' marks all duplicates as True (so keeps transcript IDs with multiple exons these)\n",
    "        mod_gr = mod_gr.subset(lambda df: df.duplicated(subset=[id_col], keep=False), nb_cpu=1)\n",
    "\n",
    "        eprint(f\"After: {len(set(mod_gr.as_df()[id_col]))}\")\n",
    "\n",
    "    # Pick last region entry by max region number for each transcript (id_col)\n",
    "    # keep=\"last\" sets last in ID to 'False' and all others true (negate to keep last only)\n",
    "\n",
    "    out_gr = mod_gr.subset(lambda df: ~(df.duplicated(subset=[id_col], keep=\"last\")),\n",
    "                           nb_cpu=1\n",
    "                           )\n",
    "\n",
    "    # re-sort by genomic coords not transcript_id\n",
    "    return out_gr.sort()\n",
    "\n",
    "\n",
    "def _df_add_region_number(df, id_col, sort_col=\"Start\"):\n",
    "    '''\n",
    "    Return a Series of strand-aware region numbers (5'-3' in 1..n)\n",
    "    Function to be used internally in a pr.assign (mainly by add_region_number)\n",
    "    '''\n",
    "    if id_col not in df.columns.tolist():\n",
    "        raise KeyError(f\"id_col - {id_col} - is not present in df for chr/strand pair {','.join([df.Chromosome.iloc[0], df.Strand.iloc[0]])}\")\n",
    "\n",
    "    elif (df.Strand == \"+\").all():\n",
    "        # Start position smallest to largest = 5'-3'\n",
    "\n",
    "        return df.groupby(id_col)[sort_col].rank(method=\"min\", ascending=True)\n",
    "\n",
    "    elif (df.Strand == \"-\").all():\n",
    "        # Start position largest to smallest = 5'-3'\n",
    "\n",
    "        return df.groupby(id_col)[sort_col].rank(method=\"min\", ascending=False)\n",
    "\n",
    "    elif df.empty:\n",
    "        eprint(\"df is empty - returning empty pd.Series\")\n",
    "        return pd.Series()\n",
    "\n",
    "\n",
    "def add_region_number(gr,\n",
    "                      id_col=\"transcript_id\",\n",
    "                      out_col=\"exon_number\",\n",
    "                      feature_col=\"Feature\",\n",
    "                      feature_key=\"exon\",\n",
    "                      ):\n",
    "    '''\n",
    "    Adds column to gr containing a strand aware region number column,\n",
    "    ordered 5'-3' 1..n by a group of features (e.g. transcript)\n",
    "    \n",
    "    Assumes every region/row of feature in group is not duplicated (i.e. given exon only appears once)\n",
    "    '''\n",
    "\n",
    "    # Make sure only 'feature_key' rows are in the gr\n",
    "    assert gr.as_df()[feature_col].drop_duplicates().tolist() == [feature_key], \"only {} entries should be present in gr\".format(feature_key)\n",
    "\n",
    "    # Make sure sorted by position first.\n",
    "    gr = gr.sort()\n",
    "\n",
    "    # Add in region number column in strand aware manner, so 1 = most 5', n = most 3'\n",
    "\n",
    "    gr = gr.assign(out_col, lambda df: _df_add_region_number(df, id_col))\n",
    "\n",
    "    return gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ede4406c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>629061</td>\n",
       "      <td>629433</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000225972.1</td>\n",
       "      <td>ENST00000416931.1</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>629639</td>\n",
       "      <td>630683</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000225630.1</td>\n",
       "      <td>ENST00000457540.1</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>631073</td>\n",
       "      <td>632616</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000237973.1</td>\n",
       "      <td>ENST00000414273.1</td>\n",
       "      <td>MTCO1P12</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>633695</td>\n",
       "      <td>634376</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000248527.1</td>\n",
       "      <td>ENST00000514057.1</td>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>634375</td>\n",
       "      <td>634922</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198744.5</td>\n",
       "      <td>ENST00000416718.2</td>\n",
       "      <td>MTCO3P12</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20503208</td>\n",
       "      <td>20507456</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229236.3</td>\n",
       "      <td>ENST00000667496.1</td>\n",
       "      <td>TTTY10</td>\n",
       "      <td>2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20504475</td>\n",
       "      <td>20507456</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229236.3</td>\n",
       "      <td>ENST00000659275.1</td>\n",
       "      <td>TTTY10</td>\n",
       "      <td>2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20506768</td>\n",
       "      <td>20507456</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229236.3</td>\n",
       "      <td>ENST00000666666.1</td>\n",
       "      <td>TTTY10</td>\n",
       "      <td>4</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20570221</td>\n",
       "      <td>20570519</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277438.1</td>\n",
       "      <td>ENST00000622542.1</td>\n",
       "      <td>KDM5DP1</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>21038288</td>\n",
       "      <td>21039044</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000254488.1</td>\n",
       "      <td>ENST00000527562.1</td>\n",
       "      <td>RP11-65G9.1</td>\n",
       "      <td>3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119902 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +7    |\n",
       "| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 629061    | 629433    | +            | ...   |\n",
       "| chr1         | exon       | 629639    | 630683    | +            | ...   |\n",
       "| chr1         | exon       | 631073    | 632616    | +            | ...   |\n",
       "| chr1         | exon       | 633695    | 634376    | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 20504475  | 20507456  | -            | ...   |\n",
       "| chrY         | exon       | 20506768  | 20507456  | -            | ...   |\n",
       "| chrY         | exon       | 20570221  | 20570519  | -            | ...   |\n",
       "| chrY         | exon       | 21038288  | 21039044  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 119,902 rows and 12 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "7 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 1 more.)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "te_mayr_cd_r3_tpm_filt = get_terminal_exons(exons_mayr_cd_r3_tpm_filt)\n",
    "\n",
    "te_mayr_cd_r3_tpm_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2db7e590",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169804074</td>\n",
       "      <td>169804386</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169807790</td>\n",
       "      <td>169807837</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169821678</td>\n",
       "      <td>169821719</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169852789</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196672977</td>\n",
       "      <td>196673407</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20570221</td>\n",
       "      <td>20570519</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277438.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11153857</td>\n",
       "      <td>11154070</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11155653</td>\n",
       "      <td>11155884</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11162449</td>\n",
       "      <td>11163137</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12661308</td>\n",
       "      <td>12662124</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000286009.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48902 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+--------------------+\n",
       "| Chromosome   | Start     | End       | Strand       | gene_id            |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)           |\n",
       "|--------------+-----------+-----------+--------------+--------------------|\n",
       "| chr1         | 169804074 | 169804386 | +            | ENSG00000000460.17 |\n",
       "| chr1         | 169807790 | 169807837 | +            | ENSG00000000460.17 |\n",
       "| chr1         | 169821678 | 169821719 | +            | ENSG00000000460.17 |\n",
       "| chr1         | 169852789 | 169854080 | +            | ENSG00000000460.17 |\n",
       "| ...          | ...       | ...       | ...          | ...                |\n",
       "| chrY         | 11153857  | 11154070  | -            | ENSG00000278212.2  |\n",
       "| chrY         | 11155653  | 11155884  | -            | ENSG00000278212.2  |\n",
       "| chrY         | 11162449  | 11163137  | -            | ENSG00000278212.2  |\n",
       "| chrY         | 12661308  | 12662124  | -            | ENSG00000286009.2  |\n",
       "+--------------+-----------+-----------+--------------+--------------------+\n",
       "Stranded PyRanges object has 48,902 rows and 5 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge TEs into non-overlapping, union TEs\n",
    "m_te_mayr_cd_r3_tpm_filt = te_mayr_cd_r3_tpm_filt.merge(strand=True, by=\"gene_id\")\n",
    "m_te_mayr_cd_r3_tpm_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1866c44f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 53.7 s, sys: 11.7 ms, total: 53.7 s\n",
      "Wall time: 53.7 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>169804074</td>\n",
       "      <td>169804386</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>ENST00000472795.5;ENST00000496973.5</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>6;6</td>\n",
       "      <td>protein_coding;protein_coding</td>\n",
       "      <td>protein_coding;protein_coding</td>\n",
       "      <td>2;1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>169807790</td>\n",
       "      <td>169807837</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>ENST00000481744.5</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>169821678</td>\n",
       "      <td>169821719</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>ENST00000466580.6</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>169852789</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>ENST00000413811.3;ENST00000498289.5;ENST000004...</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>23;29;23;24;25</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding;p...</td>\n",
       "      <td>protein_coding;processed_transcript;nonsense_m...</td>\n",
       "      <td>1;2;2;1;1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>196672977</td>\n",
       "      <td>196673407</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.16</td>\n",
       "      <td>ENST00000496761.1</td>\n",
       "      <td>CFH</td>\n",
       "      <td>2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>20570221</td>\n",
       "      <td>20570519</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277438.1</td>\n",
       "      <td>ENST00000622542.1</td>\n",
       "      <td>KDM5DP1</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "      <td>48898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>11153857</td>\n",
       "      <td>11154070</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "      <td>ENST00000620700.2</td>\n",
       "      <td>MAFIP</td>\n",
       "      <td>6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NA</td>\n",
       "      <td>48899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>11155653</td>\n",
       "      <td>11155884</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "      <td>ENST00000652671.1</td>\n",
       "      <td>MAFIP</td>\n",
       "      <td>3</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>nan</td>\n",
       "      <td>48900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>11162449</td>\n",
       "      <td>11163137</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "      <td>ENST00000651211.1</td>\n",
       "      <td>MAFIP</td>\n",
       "      <td>3</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>nan</td>\n",
       "      <td>48901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>12661308</td>\n",
       "      <td>12662124</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000286009.2</td>\n",
       "      <td>ENST00000652216.1;ENST00000662422.1;ENST000006...</td>\n",
       "      <td>RP11-125B2.1</td>\n",
       "      <td>3;3;3;3</td>\n",
       "      <td>lncRNA;lncRNA;lncRNA;lncRNA</td>\n",
       "      <td>lncRNA;lncRNA;lncRNA;lncRNA</td>\n",
       "      <td>nan;nan;nan;nan</td>\n",
       "      <td>48902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48902 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Feature    | Start     | End       | Strand       | +8    |\n",
       "| (category)   | (object)   | (int32)   | (int32)   | (category)   | ...   |\n",
       "|--------------+------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | exon       | 169804074 | 169804386 | +            | ...   |\n",
       "| chr1         | exon       | 169807790 | 169807837 | +            | ...   |\n",
       "| chr1         | exon       | 169821678 | 169821719 | +            | ...   |\n",
       "| chr1         | exon       | 169852789 | 169854080 | +            | ...   |\n",
       "| ...          | ...        | ...       | ...       | ...          | ...   |\n",
       "| chrY         | exon       | 11153857  | 11154070  | -            | ...   |\n",
       "| chrY         | exon       | 11155653  | 11155884  | -            | ...   |\n",
       "| chrY         | exon       | 11162449  | 11163137  | -            | ...   |\n",
       "| chrY         | exon       | 12661308  | 12662124  | -            | ...   |\n",
       "+--------------+------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 48,902 rows and 13 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: gene_id, transcript_id, gene_name, exon_number, gene_type, transcript_type, ... (+ 2 more.)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# If want to retain metadata (e.g. transcript IDs) for overlapping TEs, can use functions below\n",
    "# This approach is a fair bit slower so won't use ds (as just need merged regions), but relevant for Matt's suggestion of a terminal exon ID\n",
    "# could probably speed up with nb_cpu for multiple processors...\n",
    "def _collapse_tes(df,\n",
    "                  cluster_col,\n",
    "                  distinct_cols\n",
    "                  ):\n",
    "    '''\n",
    "    '''\n",
    " \n",
    "    col_order = df.columns.tolist()\n",
    "            \n",
    "    other_cols = [col for col in col_order if col not in distinct_cols]\n",
    "    # First define how want to collapse metadata columns to single row values\n",
    "    # \n",
    "    # Cols want to combine values as ; separated string (e.g. if expect diff values per row)\n",
    "    cat_dict = {col: lambda x: x.astype(str).str.cat(sep=\";\") \n",
    "                  for col in df.columns\n",
    "                    if col not in distinct_cols}\n",
    "\n",
    "    # Cols want to report a single value (e.g. where expect same values in all rows/not important)\n",
    "    # Also important where need to maintain original \n",
    "    distinct_dict = {col: lambda x: x.first()\n",
    "                         for col in distinct_cols\n",
    "                        }\n",
    "    \n",
    "    # Define the merged ranges\n",
    "    cat_dict[\"Start\"] = lambda x: x.min()\n",
    "    cat_dict[\"End\"] = lambda x: x.max()\n",
    "    \n",
    "    grpd = df.groupby(cluster_col, group_keys=False)\n",
    "    \n",
    "    grpd_distinct = grpd[distinct_cols].first()\n",
    "    grpd_other = grpd[other_cols].agg(cat_dict)\n",
    "    \n",
    "    # Combine to single row\n",
    "    clpsd = grpd_distinct.merge(grpd_other, left_index=True, right_index=True)\n",
    "    \n",
    "    return clpsd[col_order]\n",
    "\n",
    "\n",
    "def collapse_tes(gr,\n",
    "                 cluster_col=\"Cluster\",\n",
    "                 distinct_cols=[\"Chromosome\", \"Feature\", \"gene_id\", \"Strand\", \"gene_name\"]):\n",
    "    \n",
    "    # Want to keep grouping col to single value\n",
    "    dist_cols = distinct_cols + [cluster_col]\n",
    "    \n",
    "    return gr.apply(lambda df: _collapse_tes(df, cluster_col, dist_cols))\n",
    "\n",
    "# i = 0 \n",
    "# for _, df in tmp_grp:\n",
    "#     while i < 2:\n",
    "#         print(df)\n",
    "#         tmp_df = df.copy()\n",
    "#         cat_dict = {col: lambda x: x.astype(str).str.cat(sep=\";\") \n",
    "#                     for col in tmp_df.columns\n",
    "#                     if col not in [\"Chromosome\", \"Start\", \"End\", \"Feature\", \"gene_id\", \"Strand\"]}\n",
    "\n",
    "#         distinct_dict = {col: lambda x: x.head(n=1)\n",
    "#                          for col in [\"Chromosome\", \"Feature\", \"gene_id\", \"Strand\"]\n",
    "#                         }\n",
    "        \n",
    "#         agg_dict = {** cat_dict, ** distinct_dict}\n",
    "        \n",
    "#         agg_dict[\"Start\"] = lambda x: x.min()\n",
    "#         agg_dict[\"End\"] = lambda x: x.max()\n",
    "        \n",
    "#         print(df.agg(agg_dict))\n",
    "                \n",
    "#         i +=1\n",
    "#     else:\n",
    "#         break\n",
    "        \n",
    "collapse_tes(te_mayr_cd_r3_tpm_filt.cluster(strand=True, by=\"gene_id\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "222f63db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Strand  Strand_b\n",
       "+       -           799\n",
       "-       +           799\n",
       "+       +           332\n",
       "-       -           292\n",
       "dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To do/check: if we have unstranded RNA_seq data, consider removing TEs that have any overlap with TEs of diff genes?\n",
    "\n",
    "(m_te_mayr_cd_r3_tpm_filt.join(m_te_mayr_cd_r3_tpm_filt,\n",
    "                               how=\"left\",\n",
    "                               )\n",
    " .subset(lambda df: df[\"gene_id\"] != df[\"gene_id_b\"])\n",
    " .as_df()\n",
    " [[\"Strand\", \"Strand_b\"]].value_counts()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "390d47ee",
   "metadata": {},
   "source": [
    "**NB: I don't think above is massively important but keeping for completeness**\n",
    "\n",
    "A very small minority of TEs overlap with TEs of other gene_ids... Most of these overlaps occur on opposite strands, so algos that can profile txpn strand & prediction datasets that are strand-specific should be able to differentiate.\n",
    "\n",
    "TODO: check overlap of TEs with other exons (i.e. not just terminal exons). Do we want to remove? Maybe not necessary, don't know if tools try to handle this uniformly. GT is just 3'seq (mostly) so should define the PASs. May inflate the number of FNs if a tool removes these kinds of sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c80843b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169804074</td>\n",
       "      <td>169804386</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>chr1:169804074:169804386:+:ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169807790</td>\n",
       "      <td>169807837</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>chr1:169807790:169807837:+:ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169821678</td>\n",
       "      <td>169821719</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>chr1:169821678:169821719:+:ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169852789</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>chr1:169852789:169854080:+:ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>196672977</td>\n",
       "      <td>196673407</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000971.16</td>\n",
       "      <td>chr1:196672977:196673407:+:ENSG00000000971.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48897</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20570221</td>\n",
       "      <td>20570519</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277438.1</td>\n",
       "      <td>chrY:20570221:20570519:-:ENSG00000277438.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48898</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11153857</td>\n",
       "      <td>11154070</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "      <td>chrY:11153857:11154070:-:ENSG00000278212.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48899</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11155653</td>\n",
       "      <td>11155884</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "      <td>chrY:11155653:11155884:-:ENSG00000278212.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11162449</td>\n",
       "      <td>11163137</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278212.2</td>\n",
       "      <td>chrY:11162449:11163137:-:ENSG00000278212.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12661308</td>\n",
       "      <td>12662124</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000286009.2</td>\n",
       "      <td>chrY:12661308:12662124:-:ENSG00000286009.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48902 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 169804074 | 169804386 | +            | ...   |\n",
       "| chr1         | 169807790 | 169807837 | +            | ...   |\n",
       "| chr1         | 169821678 | 169821719 | +            | ...   |\n",
       "| chr1         | 169852789 | 169854080 | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 11153857  | 11154070  | -            | ...   |\n",
       "| chrY         | 11155653  | 11155884  | -            | ...   |\n",
       "| chrY         | 11162449  | 11163137  | -            | ...   |\n",
       "| chrY         | 12661308  | 12662124  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 48,902 rows and 6 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: gene_id, te_id"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def assign_id(gr, cols_to_cat=[\"Chromosome\",\"Start\",\"End\",\"Strand\"], sep_char=\":\", out_col=\"pas_id\"):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    return gr.assign(out_col,\n",
    "                     lambda df: df[cols_to_cat[0]].str.cat(df[cols_to_cat[1:]].astype(str),\n",
    "                                                           sep=sep_char\n",
    "                                                           )\n",
    "                    )\n",
    "\n",
    "m_te_mayr_cd_r3_tpm_filt = assign_id(m_te_mayr_cd_r3_tpm_filt,\n",
    "                                     cols_to_cat=[\"Chromosome\",\"Start\",\"End\",\"Strand\", \"gene_id\"],\n",
    "                                     out_col=\"te_id\")\n",
    "m_te_mayr_cd_r3_tpm_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b26eb91",
   "metadata": {},
   "source": [
    "Have a set of TEs that pass expression filter, now can filter the GT PAS for those overlapping these TEs\n",
    "\n",
    "\n",
    "# 3. Select TEs that have APA (>= 2 polyA sites)\n",
    "\n",
    "Want to also track:\n",
    "1. Number of PAS that are lost to filtering\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e21b3350",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>629238</td>\n",
       "      <td>629239</td>\n",
       "      <td>1:629219-629259:+</td>\n",
       "      <td>13.2345</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>630540</td>\n",
       "      <td>630541</td>\n",
       "      <td>1:630520-630563:+</td>\n",
       "      <td>58.5904</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>634373</td>\n",
       "      <td>634374</td>\n",
       "      <td>1:634350-634378:+</td>\n",
       "      <td>87.2652</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014536</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1:1014512-1014544:+</td>\n",
       "      <td>38.4628</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1235003</td>\n",
       "      <td>1235004</td>\n",
       "      <td>1:1235003-1235008:+</td>\n",
       "      <td>0.4136</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13338</th>\n",
       "      <td>Y</td>\n",
       "      <td>13323034</td>\n",
       "      <td>13323035</td>\n",
       "      <td>Y:13323029-13323059:-</td>\n",
       "      <td>9.5123</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13339</th>\n",
       "      <td>Y</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13340</th>\n",
       "      <td>Y</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13341</th>\n",
       "      <td>Y</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13342</th>\n",
       "      <td>Y</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13343 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| 1            | 629238    | 629239    | 1:629219-629259:+     | ...   |\n",
       "| 1            | 630540    | 630541    | 1:630520-630563:+     | ...   |\n",
       "| 1            | 634373    | 634374    | 1:634350-634378:+     | ...   |\n",
       "| 1            | 1014536   | 1014537   | 1:1014512-1014544:+   | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| Y            | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| Y            | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| Y            | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| Y            | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 13,343 rows and 6 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Score, Strand"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt_pas = pr.read_bed(\"data/gt/Mayr_CD5B_R3.SRR6795684.3seq.hg38.bed\")\n",
    "gt_pas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8d953f7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629238</td>\n",
       "      <td>629239</td>\n",
       "      <td>1:629219-629259:+</td>\n",
       "      <td>13.2345</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630540</td>\n",
       "      <td>630541</td>\n",
       "      <td>1:630520-630563:+</td>\n",
       "      <td>58.5904</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>634373</td>\n",
       "      <td>634374</td>\n",
       "      <td>1:634350-634378:+</td>\n",
       "      <td>87.2652</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014536</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1:1014512-1014544:+</td>\n",
       "      <td>38.4628</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1235003</td>\n",
       "      <td>1235004</td>\n",
       "      <td>1:1235003-1235008:+</td>\n",
       "      <td>0.4136</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13338</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13323034</td>\n",
       "      <td>13323035</td>\n",
       "      <td>Y:13323029-13323059:-</td>\n",
       "      <td>9.5123</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13339</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13340</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13341</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13342</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13343 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| chr1         | 629238    | 629239    | 1:629219-629259:+     | ...   |\n",
       "| chr1         | 630540    | 630541    | 1:630520-630563:+     | ...   |\n",
       "| chr1         | 634373    | 634374    | 1:634350-634378:+     | ...   |\n",
       "| chr1         | 1014536   | 1014537   | 1:1014512-1014544:+   | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 13,343 rows and 6 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Score, Strand"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt_pas.Chromosome = \"chr\" + gt_pas.Chromosome.astype(str)\n",
    "gt_pas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "685866f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629238</td>\n",
       "      <td>629239</td>\n",
       "      <td>1:629219-629259:+</td>\n",
       "      <td>13.2345</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:629238:629239:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630540</td>\n",
       "      <td>630541</td>\n",
       "      <td>1:630520-630563:+</td>\n",
       "      <td>58.5904</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:630540:630541:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>634373</td>\n",
       "      <td>634374</td>\n",
       "      <td>1:634350-634378:+</td>\n",
       "      <td>87.2652</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:634373:634374:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014536</td>\n",
       "      <td>1014537</td>\n",
       "      <td>1:1014512-1014544:+</td>\n",
       "      <td>38.4628</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1014536:1014537:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1235003</td>\n",
       "      <td>1235004</td>\n",
       "      <td>1:1235003-1235008:+</td>\n",
       "      <td>0.4136</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1235003:1235004:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13338</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13323034</td>\n",
       "      <td>13323035</td>\n",
       "      <td>Y:13323029-13323059:-</td>\n",
       "      <td>9.5123</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13323034:13323035:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13339</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13340</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692471:19692472:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13341</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13342</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705605:19705606:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13343 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +3    |\n",
       "| (category)   | (int32)   | (int32)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| chr1         | 629238    | 629239    | 1:629219-629259:+     | ...   |\n",
       "| chr1         | 630540    | 630541    | 1:630520-630563:+     | ...   |\n",
       "| chr1         | 634373    | 634374    | 1:634350-634378:+     | ...   |\n",
       "| chr1         | 1014536   | 1014537   | 1:1014512-1014544:+   | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 13,343 rows and 7 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Score, Strand, pas_id"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt_pas = assign_id(gt_pas)\n",
    "gt_pas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4aa9f25b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1',\n",
       " 'chr2',\n",
       " 'chr3',\n",
       " 'chr4',\n",
       " 'chr5',\n",
       " 'chr6',\n",
       " 'chr7',\n",
       " 'chr8',\n",
       " 'chr9',\n",
       " 'chr10',\n",
       " 'chr11',\n",
       " 'chr12',\n",
       " 'chr13',\n",
       " 'chr14',\n",
       " 'chr15',\n",
       " 'chr16',\n",
       " 'chr17',\n",
       " 'chr18',\n",
       " 'chr19',\n",
       " 'chr20',\n",
       " 'chr21',\n",
       " 'chr22',\n",
       " 'chrM',\n",
       " 'chrX',\n",
       " 'chrY']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_te_mayr_cd_r3_tpm_filt.Chromosome.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e262d2fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13343"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_gt_pas = gt_pas.pas_id.nunique()\n",
    "n_gt_pas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b69a00a",
   "metadata": {},
   "source": [
    "How many GT PAS do not overlap filtered terminal exons?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "22819837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5051"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "gt_pas = pr.PyRanges(gt_pas.as_df(), int64=True)\n",
    "m_te_mayr_cd_r3_tpm_filt = pr.PyRanges(m_te_mayr_cd_r3_tpm_filt.as_df(), int64=True)\n",
    "\n",
    "# pr.join often complains if the int dtypes of coordinates are different between PyRanges/not int64\n",
    "n_gt_pas_no_olap = (gt_pas.join(m_te_mayr_cd_r3_tpm_filt,\n",
    "             how=\"left\", strandedness=\"same\")\n",
    " .subset(lambda df: df[\"te_id\"] == \"-1\")\n",
    " .pas_id.nunique()\n",
    ")\n",
    "\n",
    "n_gt_pas_no_olap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "91196283",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.37855055085063327"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_gt_pas_no_olap / n_gt_pas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb90b042",
   "metadata": {},
   "source": [
    "~ 38 % do not overlap with filtered TEs...\n",
    "\n",
    "How many PAS occur on TEs by themselves? i.e. only a single GT PAS for that TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "53b97024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4730"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_gt_pas_single = (m_te_mayr_cd_r3_tpm_filt.count_overlaps(gt_pas, strandedness=\"same\")\n",
    " .subset(lambda df: df[\"NumberOverlaps\"] == 1)\n",
    " # Since each TE only has one overlapping pas, can use te_id count as proxy for pas IDs\n",
    " .te_id.nunique()\n",
    ")\n",
    "\n",
    "n_gt_pas_single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "7b0a25f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.35449299258037925"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_gt_pas_single / n_gt_pas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1a5a8fb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3562"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_gt_pas_min2 = n_gt_pas - n_gt_pas_no_olap - n_gt_pas_single\n",
    "n_gt_pas_min2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1a88324",
   "metadata": {},
   "source": [
    "Only 3562 / 13343 PAS are overlapping filtered TEs for our analysis. This number of PAS will shrink as select 2 representative PAS for TEs that have > 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "78435dd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    48902.00000\n",
       "mean         0.17077\n",
       "std          0.55754\n",
       "min          0.00000\n",
       "0%           0.00000\n",
       "5%           0.00000\n",
       "10%          0.00000\n",
       "15%          0.00000\n",
       "20%          0.00000\n",
       "25%          0.00000\n",
       "30%          0.00000\n",
       "35%          0.00000\n",
       "40%          0.00000\n",
       "45%          0.00000\n",
       "50%          0.00000\n",
       "55%          0.00000\n",
       "60%          0.00000\n",
       "65%          0.00000\n",
       "70%          0.00000\n",
       "75%          0.00000\n",
       "80%          0.00000\n",
       "85%          0.00000\n",
       "90%          1.00000\n",
       "95%          1.00000\n",
       "max         34.00000\n",
       "Name: NumberOverlaps, dtype: float64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEs vs number of overlapping PAS\n",
    "# Again perhaps not so surprising to see large propn not having APA\n",
    "## Isoform not necessarily expressed, TE only has 1 pas etc.\n",
    "(m_te_mayr_cd_r3_tpm_filt.count_overlaps(gt_pas, strandedness=\"same\")\n",
    " .as_df()[\"NumberOverlaps\"].describe(percentiles = [i * 0.01 for i in range(0,100,5)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60e47c0e",
   "metadata": {},
   "source": [
    "Now subset to valid TEs (at least two PAS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e0447a62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "      <th>NumberOverlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24468985</td>\n",
       "      <td>24472976</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>chr1:24468985:24472976:+:ENSG00000001461.17</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11840142</td>\n",
       "      <td>11843143</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>chr1:11840142:11843143:+:ENSG00000011021.23</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150158818</td>\n",
       "      <td>150160065</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000023902.14</td>\n",
       "      <td>chr1:150158818:150160065:+:ENSG00000023902.14</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12508892</td>\n",
       "      <td>12512047</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000048707.15</td>\n",
       "      <td>chr1:12508892:12512047:+:ENSG00000048707.15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16395037</td>\n",
       "      <td>16398145</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000055070.17</td>\n",
       "      <td>chr1:16395037:16398145:+:ENSG00000055070.17</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1411</th>\n",
       "      <td>chrX</td>\n",
       "      <td>72272041</td>\n",
       "      <td>72272772</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198034.11</td>\n",
       "      <td>chrX:72272041:72272772:-:ENSG00000198034.11</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149477103</td>\n",
       "      <td>149483218</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000241489.8</td>\n",
       "      <td>chrX:149477103:149483218:-:ENSG00000241489.8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19603796</td>\n",
       "      <td>19606274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000131002.12</td>\n",
       "      <td>chrY:19603796:19606274:+:ENSG00000131002.12</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1416 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 24468985  | 24472976  | +            | ...   |\n",
       "| chr1         | 11840142  | 11843143  | +            | ...   |\n",
       "| chr1         | 150158818 | 150160065 | +            | ...   |\n",
       "| chr1         | 12508892  | 12512047  | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrX         | 107713220 | 107714749 | -            | ...   |\n",
       "| chrX         | 1386151   | 1386759   | -            | ...   |\n",
       "| chrX         | 72272041  | 72272772  | -            | ...   |\n",
       "| chrX         | 149477103 | 149483218 | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 1,416 rows and 7 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: gene_id, te_id, NumberOverlaps"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEs have at least 2 overlapping PAS\n",
    "valid_tes = (m_te_mayr_cd_r3_tpm_filt.count_overlaps(gt_pas, strandedness=\"same\")\n",
    "             .subset(lambda df: df[\"NumberOverlaps\"] >= 2))\n",
    "\n",
    "valid_tes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "7536cea4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TEs with at least 2 GT PAS - 1416\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of TEs with at least 2 GT PAS - {valid_tes.te_id.nunique()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a7138568",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "      <th>NumberOverlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310137</td>\n",
       "      <td>2310138</td>\n",
       "      <td>1:2310102-2310140:+</td>\n",
       "      <td>118.9729</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310137:2310138:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310180</td>\n",
       "      <td>2310181</td>\n",
       "      <td>1:2310180-2310181:+</td>\n",
       "      <td>0.2757</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310180:2310181:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310205</td>\n",
       "      <td>2310206</td>\n",
       "      <td>1:2310183-2310220:+</td>\n",
       "      <td>13.5102</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310205:2310206:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2403215</td>\n",
       "      <td>2403216</td>\n",
       "      <td>1:2403190-2403237:+</td>\n",
       "      <td>31.7077</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2403215:2403216:+</td>\n",
       "      <td>2403034</td>\n",
       "      <td>2405442</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157916.20</td>\n",
       "      <td>chr1:2403034:2405442:+:ENSG00000157916.20</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2403757</td>\n",
       "      <td>2403758</td>\n",
       "      <td>1:2403743-2403782:+</td>\n",
       "      <td>10.6152</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2403757:2403758:+</td>\n",
       "      <td>2403034</td>\n",
       "      <td>2405442</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157916.20</td>\n",
       "      <td>chr1:2403034:2405442:+:ENSG00000157916.20</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3616</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19605813</td>\n",
       "      <td>19605814</td>\n",
       "      <td>Y:19605788-19605818:+</td>\n",
       "      <td>41.7715</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19605813:19605814:+</td>\n",
       "      <td>19603796</td>\n",
       "      <td>19606274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000131002.12</td>\n",
       "      <td>chrY:19603796:19606274:+:ENSG00000131002.12</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3617</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692471:19692472:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3619</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3620</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705605:19705606:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3621 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| chr1         | 2310137   | 2310138   | 1:2310102-2310140:+   | ...   |\n",
       "| chr1         | 2310180   | 2310181   | 1:2310180-2310181:+   | ...   |\n",
       "| chr1         | 2310205   | 2310206   | 1:2310183-2310220:+   | ...   |\n",
       "| chr1         | 2403215   | 2403216   | 1:2403190-2403237:+   | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 3,621 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 2 more.)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now subset to PAS overlapping these TEs\n",
    "gt_pas_tes = gt_pas.join(valid_tes, strandedness=\"same\")\n",
    "\n",
    "gt_pas_tes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7577424",
   "metadata": {},
   "source": [
    "\n",
    "## 4. Select two representative PAS for TEs with > 2 PASs\n",
    "\n",
    "If TE has 2 PAS - those are the selected PAS for that TE.\n",
    "All other TEs - need to select two representative PAS, Joseph's suggestion was to take the two most highly expressed.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "79c53238",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "      <th>NumberOverlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101241516</td>\n",
       "      <td>101241517</td>\n",
       "      <td>1:101241488-101241534:+</td>\n",
       "      <td>102.2919</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:101241516:101241517:+</td>\n",
       "      <td>101238821</td>\n",
       "      <td>101241518</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000170989.10</td>\n",
       "      <td>chr1:101238821:101241518:+:ENSG00000170989.10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101241381</td>\n",
       "      <td>101241382</td>\n",
       "      <td>1:101241358-101241382:+</td>\n",
       "      <td>9.6502</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:101241381:101241382:+</td>\n",
       "      <td>101238821</td>\n",
       "      <td>101241518</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000170989.10</td>\n",
       "      <td>chr1:101238821:101241518:+:ENSG00000170989.10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108700364</td>\n",
       "      <td>108700365</td>\n",
       "      <td>1:108700340-108700378:+</td>\n",
       "      <td>26.6069</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:108700364:108700365:+</td>\n",
       "      <td>108699161</td>\n",
       "      <td>108702928</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134186.12</td>\n",
       "      <td>chr1:108699161:108702928:+:ENSG00000134186.12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108701192</td>\n",
       "      <td>108701193</td>\n",
       "      <td>1:108701170-108701193:+</td>\n",
       "      <td>13.7860</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:108701192:108701193:+</td>\n",
       "      <td>108699161</td>\n",
       "      <td>108702928</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134186.12</td>\n",
       "      <td>chr1:108699161:108702928:+:ENSG00000134186.12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110899919</td>\n",
       "      <td>110899920</td>\n",
       "      <td>1:110899894-110899942:+</td>\n",
       "      <td>820.5406</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:110899919:110899920:+</td>\n",
       "      <td>110899123</td>\n",
       "      <td>110899922</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000143119.14</td>\n",
       "      <td>chr1:110899123:110899922:+:ENSG00000143119.14</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>chrX</td>\n",
       "      <td>78961834</td>\n",
       "      <td>78961835</td>\n",
       "      <td>X:78961813-78961835:+</td>\n",
       "      <td>10.2016</td>\n",
       "      <td>+</td>\n",
       "      <td>chrX:78961834:78961835:+</td>\n",
       "      <td>78960507</td>\n",
       "      <td>78963727</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000078589.13</td>\n",
       "      <td>chrX:78960507:78963727:+:ENSG00000078589.13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>chrX</td>\n",
       "      <td>107713326</td>\n",
       "      <td>107713327</td>\n",
       "      <td>X:107713301-107713352:-</td>\n",
       "      <td>19.8518</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:107713326:107713327:-</td>\n",
       "      <td>107713220</td>\n",
       "      <td>107714749</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000157514.17</td>\n",
       "      <td>chrX:107713220:107714749:-:ENSG00000157514.17</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td>chrX</td>\n",
       "      <td>107713236</td>\n",
       "      <td>107713237</td>\n",
       "      <td>X:107713236-107713250:-</td>\n",
       "      <td>11.1666</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:107713236:107713237:-</td>\n",
       "      <td>107713220</td>\n",
       "      <td>107714749</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000157514.17</td>\n",
       "      <td>chrX:107713220:107714749:-:ENSG00000157514.17</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td>chrX</td>\n",
       "      <td>1386250</td>\n",
       "      <td>1386251</td>\n",
       "      <td>X:1386225-1386271:-</td>\n",
       "      <td>212.5796</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:1386250:1386251:-</td>\n",
       "      <td>1386151</td>\n",
       "      <td>1386759</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000169100.14</td>\n",
       "      <td>chrX:1386151:1386759:-:ENSG00000169100.14</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td>chrX</td>\n",
       "      <td>1386151</td>\n",
       "      <td>1386152</td>\n",
       "      <td>X:1386124-1386172:-</td>\n",
       "      <td>43.0122</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:1386151:1386152:-</td>\n",
       "      <td>1386151</td>\n",
       "      <td>1386759</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000169100.14</td>\n",
       "      <td>chrX:1386151:1386759:-:ENSG00000169100.14</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>886 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                    | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (object)                | ...   |\n",
       "|--------------+-----------+-----------+-------------------------+-------|\n",
       "| chr1         | 101241516 | 101241517 | 1:101241488-101241534:+ | ...   |\n",
       "| chr1         | 101241381 | 101241382 | 1:101241358-101241382:+ | ...   |\n",
       "| chr1         | 108700364 | 108700365 | 1:108700340-108700378:+ | ...   |\n",
       "| chr1         | 108701192 | 108701193 | 1:108701170-108701193:+ | ...   |\n",
       "| ...          | ...       | ...       | ...                     | ...   |\n",
       "| chrX         | 107713326 | 107713327 | X:107713301-107713352:- | ...   |\n",
       "| chrX         | 107713236 | 107713237 | X:107713236-107713250:- | ...   |\n",
       "| chrX         | 1386250   | 1386251   | X:1386225-1386271:-     | ...   |\n",
       "| chrX         | 1386151   | 1386152   | X:1386124-1386172:-     | ...   |\n",
       "+--------------+-----------+-----------+-------------------------+-------+\n",
       "Stranded PyRanges object has 886 rows and 13 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, te_id, ... (+ 1 more.)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apparently the rank approach is much faster if we're worried about that...\n",
    "# https://stackoverflow.com/questions/62157558/filter-for-rows-with-n-largest-values-for-each-group\n",
    "# Score is the TPM expression of a PAS\n",
    "# Select two highest expressed PASs for genes \n",
    "\n",
    "gt_pas_tes_2rep = (gt_pas_tes.subset(lambda df: df[\"NumberOverlaps\"] > 2)\n",
    " .apply(lambda df: df.groupby(\"te_id\").apply(lambda x: x.nlargest(2, \"Score\")).reset_index(drop=True))\n",
    ")\n",
    "\n",
    "gt_pas_tes_2rep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "0842e97d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "      <th>NumberOverlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11843045</td>\n",
       "      <td>11843046</td>\n",
       "      <td>1:11843016-11843046:+</td>\n",
       "      <td>13.0967</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:11843045:11843046:+</td>\n",
       "      <td>11840142</td>\n",
       "      <td>11843143</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>chr1:11840142:11843143:+:ENSG00000011021.23</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11843127</td>\n",
       "      <td>11843128</td>\n",
       "      <td>1:11843102-11843143:+</td>\n",
       "      <td>48.8023</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:11843127:11843128:+</td>\n",
       "      <td>11840142</td>\n",
       "      <td>11843143</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>chr1:11840142:11843143:+:ENSG00000011021.23</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12013470</td>\n",
       "      <td>12013471</td>\n",
       "      <td>1:12013468-12013476:+</td>\n",
       "      <td>17.2325</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:12013470:12013471:+</td>\n",
       "      <td>12011495</td>\n",
       "      <td>12013514</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116688.18</td>\n",
       "      <td>chr1:12011495:12013514:+:ENSG00000116688.18</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12013501</td>\n",
       "      <td>12013502</td>\n",
       "      <td>1:12013488-12013523:+</td>\n",
       "      <td>20.5411</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:12013501:12013502:+</td>\n",
       "      <td>12011495</td>\n",
       "      <td>12013514</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000116688.18</td>\n",
       "      <td>chr1:12011495:12013514:+:ENSG00000116688.18</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12510360</td>\n",
       "      <td>12510361</td>\n",
       "      <td>1:12510335-12510378:+</td>\n",
       "      <td>19.1625</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:12510360:12510361:+</td>\n",
       "      <td>12508892</td>\n",
       "      <td>12512047</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000048707.15</td>\n",
       "      <td>chr1:12508892:12512047:+:ENSG00000048707.15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19605813</td>\n",
       "      <td>19605814</td>\n",
       "      <td>Y:19605788-19605818:+</td>\n",
       "      <td>41.7715</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19605813:19605814:+</td>\n",
       "      <td>19603796</td>\n",
       "      <td>19606274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000131002.12</td>\n",
       "      <td>chrY:19603796:19606274:+:ENSG00000131002.12</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2828</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692471:19692472:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2830</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2831</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705605:19705606:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2832 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +9    |\n",
       "| (category)   | (int64)   | (int64)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| chr1         | 11843045  | 11843046  | 1:11843016-11843046:+ | ...   |\n",
       "| chr1         | 11843127  | 11843128  | 1:11843102-11843143:+ | ...   |\n",
       "| chr1         | 12013470  | 12013471  | 1:12013468-12013476:+ | ...   |\n",
       "| chr1         | 12013501  | 12013502  | 1:12013488-12013523:+ | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 2,832 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "9 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 2 more.)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine selected 2 with other PAS on TEs with only 2 PAS to get final list of GT PAS\n",
    "gt_pas_tes_rep = pr.concat([gt_pas_tes.subset(lambda df: df[\"NumberOverlaps\"] == 2), gt_pas_tes_2rep])\n",
    "gt_pas_tes_rep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2be19e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2818"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_gt_pas_rep = gt_pas_tes_rep.pas_id.nunique()\n",
    "n_gt_pas_rep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044961b5",
   "metadata": {},
   "source": [
    "Should be 2 PAS for every TE, but 1416 * 2 != 2818..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "dcedbdcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "      <th>NumberOverlaps</th>\n",
       "      <th>pas_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310137</td>\n",
       "      <td>2310138</td>\n",
       "      <td>1:2310102-2310140:+</td>\n",
       "      <td>118.9729</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310137:2310138:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310205</td>\n",
       "      <td>2310206</td>\n",
       "      <td>1:2310183-2310220:+</td>\n",
       "      <td>13.5102</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310205:2310206:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2403215</td>\n",
       "      <td>2403216</td>\n",
       "      <td>1:2403190-2403237:+</td>\n",
       "      <td>31.7077</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2403215:2403216:+</td>\n",
       "      <td>2403034</td>\n",
       "      <td>2405442</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157916.20</td>\n",
       "      <td>chr1:2403034:2405442:+:ENSG00000157916.20</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2405434</td>\n",
       "      <td>2405435</td>\n",
       "      <td>1:2405417-2405448:+</td>\n",
       "      <td>13.2345</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2405434:2405435:+</td>\n",
       "      <td>2403034</td>\n",
       "      <td>2405442</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157916.20</td>\n",
       "      <td>chr1:2403034:2405442:+:ENSG00000157916.20</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11843045</td>\n",
       "      <td>11843046</td>\n",
       "      <td>1:11843016-11843046:+</td>\n",
       "      <td>13.0967</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:11843045:11843046:+</td>\n",
       "      <td>11840142</td>\n",
       "      <td>11843143</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>chr1:11840142:11843143:+:ENSG00000011021.23</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19605813</td>\n",
       "      <td>19605814</td>\n",
       "      <td>Y:19605788-19605818:+</td>\n",
       "      <td>41.7715</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19605813:19605814:+</td>\n",
       "      <td>19603796</td>\n",
       "      <td>19606274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000131002.12</td>\n",
       "      <td>chrY:19603796:19606274:+:ENSG00000131002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2828</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692471:19692472:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2830</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2831</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705605:19705606:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2832 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +10   |\n",
       "| (category)   | (int64)   | (int64)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| chr1         | 2310137   | 2310138   | 1:2310102-2310140:+   | ...   |\n",
       "| chr1         | 2310205   | 2310206   | 1:2310183-2310220:+   | ...   |\n",
       "| chr1         | 2403215   | 2403216   | 1:2403190-2403237:+   | ...   |\n",
       "| chr1         | 2405434   | 2405435   | 1:2405417-2405448:+   | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 2,832 rows and 14 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "10 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 3 more.)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Label as proximal / distal - assign region number 1 (proximal) & 2 (distal)\n",
    "gt_pas_tes_rep = add_region_number(gt_pas_tes_rep.assign(\"Feature\",\n",
    "                                                         lambda x: pd.Series([\"pas\"]*len(x.index))),\n",
    "                                   id_col=\"te_id\",\n",
    "                                   out_col=\"pas_number\",\n",
    "                                   feature_key=\"pas\").drop(\"Feature\")\n",
    "\n",
    "gt_pas_tes_rep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "1acf6845",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>te_id</th>\n",
       "      <th>NumberOverlaps</th>\n",
       "      <th>pas_number</th>\n",
       "      <th>rel_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310137</td>\n",
       "      <td>2310138</td>\n",
       "      <td>1:2310102-2310140:+</td>\n",
       "      <td>118.9729</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310137:2310138:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.898023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2310205</td>\n",
       "      <td>2310206</td>\n",
       "      <td>1:2310183-2310220:+</td>\n",
       "      <td>13.5102</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2310205:2310206:+</td>\n",
       "      <td>2306576</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933.10</td>\n",
       "      <td>chr1:2306576:2310213:+:ENSG00000157933.10</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.101977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2403215</td>\n",
       "      <td>2403216</td>\n",
       "      <td>1:2403190-2403237:+</td>\n",
       "      <td>31.7077</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2403215:2403216:+</td>\n",
       "      <td>2403034</td>\n",
       "      <td>2405442</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157916.20</td>\n",
       "      <td>chr1:2403034:2405442:+:ENSG00000157916.20</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.705522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2405434</td>\n",
       "      <td>2405435</td>\n",
       "      <td>1:2405417-2405448:+</td>\n",
       "      <td>13.2345</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:2405434:2405435:+</td>\n",
       "      <td>2403034</td>\n",
       "      <td>2405442</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157916.20</td>\n",
       "      <td>chr1:2403034:2405442:+:ENSG00000157916.20</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.294478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11843045</td>\n",
       "      <td>11843046</td>\n",
       "      <td>1:11843016-11843046:+</td>\n",
       "      <td>13.0967</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:11843045:11843046:+</td>\n",
       "      <td>11840142</td>\n",
       "      <td>11843143</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000011021.23</td>\n",
       "      <td>chr1:11840142:11843143:+:ENSG00000011021.23</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.211582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19605813</td>\n",
       "      <td>19605814</td>\n",
       "      <td>Y:19605788-19605818:+</td>\n",
       "      <td>41.7715</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY:19605813:19605814:+</td>\n",
       "      <td>19603796</td>\n",
       "      <td>19606274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000131002.12</td>\n",
       "      <td>chrY:19603796:19606274:+:ENSG00000131002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.780928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2828</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19691944</td>\n",
       "      <td>19691945</td>\n",
       "      <td>Y:19691920-19691966:-</td>\n",
       "      <td>11.0288</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19691944:19691945:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.567378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19692471</td>\n",
       "      <td>19692472</td>\n",
       "      <td>Y:19692471-19692501:-</td>\n",
       "      <td>8.4094</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19692471:19692472:-</td>\n",
       "      <td>19691940</td>\n",
       "      <td>19694606</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260197.1</td>\n",
       "      <td>chrY:19691940:19694606:-:ENSG00000260197.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.432622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2830</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705419</td>\n",
       "      <td>19705420</td>\n",
       "      <td>Y:19705410-19705440:-</td>\n",
       "      <td>153.7135</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705419:19705420:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.936188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2831</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705605</td>\n",
       "      <td>19705606</td>\n",
       "      <td>Y:19705585-19705629:-</td>\n",
       "      <td>10.4773</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705605:19705606:-</td>\n",
       "      <td>19703864</td>\n",
       "      <td>19706345</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817.16</td>\n",
       "      <td>chrY:19703864:19706345:-:ENSG00000012817.16</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.063812</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2832 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                  | +11   |\n",
       "| (category)   | (int64)   | (int64)   | (object)              | ...   |\n",
       "|--------------+-----------+-----------+-----------------------+-------|\n",
       "| chr1         | 2310137   | 2310138   | 1:2310102-2310140:+   | ...   |\n",
       "| chr1         | 2310205   | 2310206   | 1:2310183-2310220:+   | ...   |\n",
       "| chr1         | 2403215   | 2403216   | 1:2403190-2403237:+   | ...   |\n",
       "| chr1         | 2405434   | 2405435   | 1:2405417-2405448:+   | ...   |\n",
       "| ...          | ...       | ...       | ...                   | ...   |\n",
       "| chrY         | 19691944  | 19691945  | Y:19691920-19691966:- | ...   |\n",
       "| chrY         | 19692471  | 19692472  | Y:19692471-19692501:- | ...   |\n",
       "| chrY         | 19705419  | 19705420  | Y:19705410-19705440:- | ...   |\n",
       "| chrY         | 19705605  | 19705606  | Y:19705585-19705629:- | ...   |\n",
       "+--------------+-----------+-----------+-----------------------+-------+\n",
       "Stranded PyRanges object has 2,832 rows and 15 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: Score, Strand, pas_id, Start_b, End_b, Strand_b, gene_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate fractional relative expression for each PAS (vs all PAS on TE)\n",
    "gt_pas_tes_rep = gt_pas_tes_rep.assign(\"rel_exp\",\n",
    "                                       lambda df: df.groupby(\"te_id\", group_keys=False)[\"Score\"].apply(lambda x: x / x.sum())\n",
    "                                       )\n",
    "\n",
    "gt_pas_tes_rep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "429bc22f",
   "metadata": {},
   "source": [
    "# Match predictions to filtered ground truth\n",
    "\n",
    "- Need to assign pred PAS to the correct TE - info not present in prediction BED files\n",
    "\n",
    "**Really need to tackle problem of multiple predictions matching GT  here (due to tools running tx-by-tx)**\n",
    "\n",
    "My initial suggestion was to take the two sites that have the 'best match' to ground truth. i.e. Take all PAS assigned to that composite/merged TE, and select the two isoforms that have the best match to the GT, in order of (both match, one of proximal/distal matches, no matches) With way I've named DaPars2 Name field there isn't an out of the box way to know which predictions have come from the same TE frame (would just have to extract the transcript ID). Since DaPars tools calc rel expression tx by tx, there could be a case where the proximal and distal sites match between different transcript models. This could mean that the predicted relative expressions do not sum to 1.\n",
    "\n",
    "\n",
    "Could probably get around this by deciding on one of prox/distal relative expr to evaluate/compare against. In this case if have multiple predicted pas in a given window, could either take the site with the smallest error/difference to GT relative expression, or take sth like the mean/median of the predicted matched sites rel exp. If did it this way, would maybe be good to try both ways around (i.e. calc metrics relative to distal rel usage and then proximal rel usage).\n",
    "         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "8a5eb89e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28453784</td>\n",
       "      <td>28453785</td>\n",
       "      <td>ENST00000438601.2_RP1-308E4.1_proximal</td>\n",
       "      <td>0.08</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:28453784:28453785:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51788223</td>\n",
       "      <td>51788224</td>\n",
       "      <td>ENST00000371714.5_OSBPL9_proximal</td>\n",
       "      <td>0.27</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:51788223:51788224:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>212103089</td>\n",
       "      <td>212103090</td>\n",
       "      <td>ENST00000475419.5_DTL_proximal</td>\n",
       "      <td>0.62</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:212103089:212103090:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>202439280</td>\n",
       "      <td>202439281</td>\n",
       "      <td>ENST00000417053.2_RP11-175B9.2_proximal</td>\n",
       "      <td>0.04</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:202439280:202439281:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24671536</td>\n",
       "      <td>24671537</td>\n",
       "      <td>ENST00000479034.5_SRRM1_proximal</td>\n",
       "      <td>0.35</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:24671536:24671537:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109649</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234579</td>\n",
       "      <td>13234580</td>\n",
       "      <td>ENST00000684226.1_UTY_distal</td>\n",
       "      <td>0.67</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234579:13234580:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109650</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2881682</td>\n",
       "      <td>2881683</td>\n",
       "      <td>ENST00000444242.1_HSFY3P_distal</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2881682:2881683:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109651</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19707181</td>\n",
       "      <td>19707182</td>\n",
       "      <td>ENST00000415360.1_KDM5D_distal</td>\n",
       "      <td>0.26</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19707181:19707182:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109652</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234580</td>\n",
       "      <td>13234581</td>\n",
       "      <td>ENST00000684326.1_UTY_distal</td>\n",
       "      <td>0.67</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234580:13234581:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109653</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705424</td>\n",
       "      <td>19705425</td>\n",
       "      <td>ENST00000492117.1_KDM5D_distal</td>\n",
       "      <td>0.73</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705424:19705425:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>109654 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +4    |\n",
       "| (category)   | (int32)   | (int32)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 28453784  | 28453785  | ...   |\n",
       "| chr1         | 51788223  | 51788224  | ...   |\n",
       "| chr1         | 212103089 | 212103090 | ...   |\n",
       "| chr1         | 202439280 | 202439281 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrY         | 2881682   | 2881683   | ...   |\n",
       "| chrY         | 19707181  | 19707182  | ...   |\n",
       "| chrY         | 13234580  | 13234581  | ...   |\n",
       "| chrY         | 19705424  | 19705425  | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 109,654 rows and 7 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Name, Score, Strand, pas_id"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_mayr_cd5_dapars2 = pr.read_bed(\"data/pred/Mayr_CD5B_R3_DaPars2_04.bed\")\n",
    "pred_mayr_cd5_dapars = pr.read_bed(\"data/pred/Mayr_CD5B_R3_relative_usage_quantification_output.bed\")\n",
    "\n",
    "\n",
    "# Name is not necessarily unique,to be safe (if need it) assign a ID for each PAS\n",
    "pred_mayr_cd5_dapars2 = assign_id(pred_mayr_cd5_dapars2)\n",
    "pred_mayr_cd5_dapars = assign_id(pred_mayr_cd5_dapars)\n",
    "\n",
    "pred_mayr_cd5_dapars2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "cd507bb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>151034007</td>\n",
       "      <td>151034008</td>\n",
       "      <td>ENST00000271620.8|chr1|151033806-151035713|+</td>\n",
       "      <td>0.30</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:151034007:151034008:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6092096</td>\n",
       "      <td>6092097</td>\n",
       "      <td>ENST00000656607.1|chr1|6091863-6092187|+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:6092096:6092097:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>156800811</td>\n",
       "      <td>156800812</td>\n",
       "      <td>ENST00000469071.1|chr1|156800374-156800812|+</td>\n",
       "      <td>0.56</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:156800811:156800812:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>111766684</td>\n",
       "      <td>111766685</td>\n",
       "      <td>ENST00000680983.1|chr1|111765737-111767243|+</td>\n",
       "      <td>0.26</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:111766684:111766685:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>32652335</td>\n",
       "      <td>32652336</td>\n",
       "      <td>ENST00000465780.1|chr1|32651914-32652506|+</td>\n",
       "      <td>0.00</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:32652335:32652336:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74179</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2883206</td>\n",
       "      <td>2883207</td>\n",
       "      <td>ENST00000652562.1|chrY|2883110-2883407|-</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2883206:2883207:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74180</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705682</td>\n",
       "      <td>19705683</td>\n",
       "      <td>ENST00000382806.6|chrY|19705420-19706345|-</td>\n",
       "      <td>0.18</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19705682:19705683:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74181</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2883109</td>\n",
       "      <td>2883110</td>\n",
       "      <td>ENST00000652562.1|chrY|2883110-2883407|-</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2883109:2883110:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74182</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13323167</td>\n",
       "      <td>13323168</td>\n",
       "      <td>ENST00000329134.9|chrY|13323034-13323368|-</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13323167:13323168:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74183</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2881682</td>\n",
       "      <td>2881683</td>\n",
       "      <td>ENST00000444242.1|chrY|2881683-2882361|-</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2881682:2881683:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74184 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +4    |\n",
       "| (category)   | (int32)   | (int32)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 151034007 | 151034008 | ...   |\n",
       "| chr1         | 6092096   | 6092097   | ...   |\n",
       "| chr1         | 156800811 | 156800812 | ...   |\n",
       "| chr1         | 111766684 | 111766685 | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrY         | 19705682  | 19705683  | ...   |\n",
       "| chrY         | 2883109   | 2883110   | ...   |\n",
       "| chrY         | 13323167  | 13323168  | ...   |\n",
       "| chrY         | 2881682   | 2881683   | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 74,184 rows and 7 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: Name, Score, Strand, pas_id"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_mayr_cd5_dapars"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9b69006",
   "metadata": {},
   "source": [
    "As a dirty readout of potential amount of duplication of PAS due to multiple TE isoforms, see how many times identical pas_ids occur\n",
    "\n",
    "Of course pas could differ very slightly if specific coords of TE isoforms are different & essentially be the same PAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "42a5b5ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of predicted pas_id that are duplicated (i.e. identical coordinates)\n",
      "DaPars\n",
      "0.09987056774420351\n",
      "DaPars2\n",
      "0.1019118652900368\n"
     ]
    }
   ],
   "source": [
    "print(\"Fraction of predicted pas_id that are duplicated (i.e. identical coordinates)\")\n",
    "print(\"DaPars\")\n",
    "print(len(pred_mayr_cd5_dapars.pas_id.value_counts().loc[lambda x: x > 1]) / pred_mayr_cd5_dapars.pas_id.nunique())\n",
    "print(\"DaPars2\")\n",
    "print(len(pred_mayr_cd5_dapars2.pas_id.value_counts().loc[lambda x: x > 1]) / pred_mayr_cd5_dapars2.pas_id.nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3679d61",
   "metadata": {},
   "source": [
    "~ 10 % of polyA sites are identical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "fa220ad6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    247\n",
      "Name: NumberOverlaps_pred, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    149\n",
       "Name: NumberOverlaps_pred, dtype: int64"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From a TE perspective - how many have >2 overlapping prediction sites?\n",
    "# In the case of DaPars, this could mean overlapping terminal exon isoforms between different transcript IDs\n",
    "##  DaPars & DaPars2 use a 2-site model, so number of predicted sites can never exceed expected 2\n",
    "# how many have no overlapping predictions (or )\n",
    "\n",
    "# Need to convert back to int32 otherwise get following error...\n",
    "# ValueError: Buffer dtype mismatch, expected 'const int32_t' but got 'long\n",
    "\n",
    "temp_dapars = pr.PyRanges(valid_tes.as_df(), int64=False).count_overlaps(pr.PyRanges(pred_mayr_cd5_dapars.as_df(), int64=False),\n",
    "                         strandedness=\"same\",\n",
    "                         keep_nonoverlapping=True,\n",
    "                         overlap_col=\"NumberOverlaps_pred\")\n",
    "temp_dapars2 = pr.PyRanges(valid_tes.as_df(), int64=False).count_overlaps(pr.PyRanges(pred_mayr_cd5_dapars2.as_df(), int64=False),\n",
    "                         strandedness=\"same\",\n",
    "                         keep_nonoverlapping=True,\n",
    "                         overlap_col=\"NumberOverlaps_pred\")\n",
    "\n",
    "\n",
    "print(temp_dapars.subset(lambda df: df[\"NumberOverlaps_pred\"] < 2).NumberOverlaps_pred.value_counts())\n",
    "temp_dapars2.subset(lambda df: df[\"NumberOverlaps_pred\"] < 2).NumberOverlaps_pred.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a68a507f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of curated GT terminal exons that have no predicted APA (i.e. < 2 PAS)\n",
      "DaPars\n",
      "0.17443502824858756\n",
      "DaPars2\n",
      "0.10522598870056497\n"
     ]
    }
   ],
   "source": [
    "print(\"Fraction of curated GT terminal exons that have no predicted APA (i.e. < 2 PAS)\")\n",
    "print(\"DaPars\")\n",
    "print(len(temp_dapars.subset(lambda df: df[\"NumberOverlaps_pred\"] < 2).NumberOverlaps_pred) / temp_dapars.te_id.nunique())\n",
    "print(\"DaPars2\")\n",
    "print(len(temp_dapars2.subset(lambda df: df[\"NumberOverlaps_pred\"] < 2).NumberOverlaps_pred) / temp_dapars2.te_id.nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016a7053",
   "metadata": {},
   "source": [
    "Above is Joseph's suggestion for 'FNs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3f26f903",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>pas_id</th>\n",
       "      <th>te_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28496684</td>\n",
       "      <td>28496685</td>\n",
       "      <td>ENST00000493669.2_PHACTR4_proximal</td>\n",
       "      <td>0.00</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:28496684:28496685:+</td>\n",
       "      <td>chr1:28496533:28500364:+:ENSG00000204138.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>53048234</td>\n",
       "      <td>53048235</td>\n",
       "      <td>ENST00000478274.6_SCP2_proximal</td>\n",
       "      <td>0.00</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:53048234:53048235:+</td>\n",
       "      <td>chr1:53047857:53051698:+:ENSG00000116171.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161120697</td>\n",
       "      <td>161120698</td>\n",
       "      <td>ENST00000368009.7_NIT1_proximal</td>\n",
       "      <td>0.48</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:161120697:161120698:+</td>\n",
       "      <td>chr1:161120498:161121194:+:ENSG00000158793.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>44778726</td>\n",
       "      <td>44778727</td>\n",
       "      <td>ENST00000396651.8_RPS8_proximal</td>\n",
       "      <td>0.11</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:44778726:44778727:+</td>\n",
       "      <td>chr1:44777999:44778779:+:ENSG00000142937.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>75762850</td>\n",
       "      <td>75762851</td>\n",
       "      <td>ENST00000473018.3_ACADM_proximal</td>\n",
       "      <td>0.48</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:75762850:75762851:+</td>\n",
       "      <td>chr1:75761121:75763720:+:ENSG00000117054.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109741</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1431901</td>\n",
       "      <td>1431902</td>\n",
       "      <td>ENST00000462195.6_PAR_Y_ASMTL_distal</td>\n",
       "      <td>0.90</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:1431901:1431902:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109742</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234579</td>\n",
       "      <td>13234580</td>\n",
       "      <td>ENST00000684226.1_UTY_distal</td>\n",
       "      <td>0.67</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234579:13234580:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109743</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2881682</td>\n",
       "      <td>2881683</td>\n",
       "      <td>ENST00000444242.1_HSFY3P_distal</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:2881682:2881683:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109744</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19707181</td>\n",
       "      <td>19707182</td>\n",
       "      <td>ENST00000415360.1_KDM5D_distal</td>\n",
       "      <td>0.26</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:19707181:19707182:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109745</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234580</td>\n",
       "      <td>13234581</td>\n",
       "      <td>ENST00000684326.1_UTY_distal</td>\n",
       "      <td>0.67</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY:13234580:13234581:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>109746 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +5    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 28496684  | 28496685  | ...   |\n",
       "| chr1         | 53048234  | 53048235  | ...   |\n",
       "| chr1         | 161120697 | 161120698 | ...   |\n",
       "| chr1         | 44778726  | 44778727  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrY         | 13234579  | 13234580  | ...   |\n",
       "| chrY         | 2881682   | 2881683   | ...   |\n",
       "| chrY         | 19707181  | 19707182  | ...   |\n",
       "| chrY         | 13234580  | 13234581  | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 109,746 rows and 8 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: Name, Score, Strand, pas_id, te_id"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign predicted PAS to GT TEs\n",
    "keep_cols = pred_mayr_cd5_dapars.columns.tolist() + [\"te_id\"]\n",
    "pred_mayr_cd5_dapars = (pr.PyRanges(pred_mayr_cd5_dapars.as_df(),int64=True).join(pr.PyRanges(valid_tes.as_df(),int64=True),\n",
    "                                                          how=\"left\",\n",
    "                                                          strandedness=\"same\", \n",
    "                                                          suffix=\"_te\"))[keep_cols]\n",
    "\n",
    "pred_mayr_cd5_dapars2 = (pr.PyRanges(pred_mayr_cd5_dapars2.as_df(),int64=True).join(pr.PyRanges(valid_tes.as_df(),\n",
    "                                                                                                int64=True),\n",
    "                                                          how=\"left\",\n",
    "                                                          strandedness=\"same\", \n",
    "                                                          suffix=\"_te\"))[keep_cols]\n",
    "\n",
    "pred_mayr_cd5_dapars2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "3215727b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of predicted PAS that do not overlap with curated terminal exons\n",
      "DaPars\n",
      "0.8895191831386523\n",
      "DaPars2\n",
      "0.911604729179559\n"
     ]
    }
   ],
   "source": [
    "print(\"Fraction of predicted PAS that do not overlap with curated terminal exons\")\n",
    "print(\"DaPars\")\n",
    "print(pred_mayr_cd5_dapars.subset(lambda df: df[\"te_id\"] == \"-1\").pas_id.nunique() / pred_mayr_cd5_dapars.pas_id.nunique())\n",
    "print(\"DaPars2\")\n",
    "print(pred_mayr_cd5_dapars2.subset(lambda df: df[\"te_id\"] == \"-1\").pas_id.nunique() / pred_mayr_cd5_dapars2.pas_id.nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0c3e8c6",
   "metadata": {},
   "source": [
    "Taking this approach of curating GT data means excluding vast majority of predicted PAS for both DaPars tools. If we considered predictions within a window to be the same PAS, that may reduce the number\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:apaeval_poc_rel_quant] *",
   "language": "python",
   "name": "conda-env-apaeval_poc_rel_quant-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
